### ISSUE TITLE
The forward function of a custom NN module in this LM method must return a 1D tensor?

### ISSUE THREAD

**Issue by nengqian11** [fabiodimarco/torch-levenberg-marquardt#6]
Hi!
Finally got some time to try this again. And I have the question as the title.
 Should the forward function of a custom NN module in this LM method must return a 1D tensor?  
For example, if the output of my module originally is a tenosr which include 2 rotation matrix ( each rotation matrix is a 3*3 matrix), then its shape will be [2,3,3].
In this case, should I transfer this [2,3,3] matrix to a [18] matrix in the forward method?

Thanks a lot!

**Comment by fabiodimarco** [comment]
@nengqian11 the output of your model can be whatever shape. But of course it needs to be coherent with the target shape. At the least all the loss function I have implemented assume the output of the network and the target have the same shape. But you can also implement a custom loss function to handle particular situations.

My implementation should offer the same flexibility you would have by using a standard optimizer provided by pytorch (Adam, SGD, etc.)

If you provide me some code examples and explanation, I can help you in formulating your problem so that it can be optimized using LM or other optimization algorithms.

**Comment by nengqian11** [comment]
Hi thanks for your reply!
 my current code is quite buggy and cannot show you.
Here is a very simple example about what I am tring to do.


```
# -----original code, 
# init some parameters here
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)
for epoch in range(max_epochs):
    optimizer.zero_grad()
    
    # some my computaion
    my_computation()

    # compute the error here, they have different shape
    some_error_1 = torch.tensor([10, 6, 3,3])   # just to show the shape
    some_error_2 = torch.tensor([10, 6, 3])     # just to show the shape
    some_loss_1 = torch.mean(some_error_1 ** 2)
    some_loss_2 = torch.mean(some_error_2 ** 2)
    loss = some_loss_1 + some_loss_2  # a scalar, can directly do backward() update.
    
    loss.backward()
    optimizer.step()


# ----- try to use the torch LM, so I defined this myModel from nn.Module
class myModel(nn.Module):
    def __init__(self) -> None:
        super().__init__()
        # some parameters init here

    def forward(self, _=None) -> torch.Tensor:
        # some my computaion
        my_computation()

        # compute the error here, they have different shape
        some_error_1 = torch.tensor([10, 6, 3,3])   # just to show the shape
        some_error_2 = torch.tensor([10, 6, 3])     # just to show the shape
        some_loss_1 = torch.mean(some_error_1 ** 2)
        some_loss_2 = torch.mean(some_error_2 ** 2)
        loss = some_loss_1 + some_loss_2
        return loss  # this is a scalar
        # or 
        # return torch.cat((some_loss_1.view(-1), some_loss_2.view(-1)))  # this is a vector

```

So originally I am using adam to do the optimization, and I do not even have a nn.Module class. I just compute all the error and then add together to the loss scalar, and do backward update.
This adam version works fine, I get the correct result, it is just slow.

So my question is how to actually convert this original code to be optimized by the torch LM optimizer?
I tried both way, seems neither is correct. The jacobian I get is all zero. 
```
  return loss  # this is a scalar
        # or 
  # return torch.cat((some_loss_1.view(-1), some_loss_2.view(-1)))  # this is a vector
```

**Comment by nengqian11** [comment]
I think I find the reason.  Why the jacobian is not correct.
Here is a new simple example.
I guess I just cannot put my parameter to a list, and then use this list to do my following forward computation.

But I do not know why? 
self.my_param_1 actually has the same ID as self.optimize_vars[0].

```
class myModel(nn.Module):
    def __init__(self )->None:
        super().__init__()
        self.my_param_1 = torch.nn.Parameter(torch.randn(4))
        self.my_param_2 = torch.nn.Parameter(torch.randn(4))

        self.param_1_target= torch.randn(4)
        self.param_2_target= torch.randn(4)
        
        self.optimize_vars = [self.my_param_1, self.my_param_2]
        
    def forward(self, _=None):
        # ---- jacobian is all zero ---------
        # error_1 = ( self.optimize_vars[0] - self.param_1_target)
        # error_2 = ( self.optimize_vars[1] - self.param_2_target)

        # ---- jacobian is correct ---------
        error_1 = (self.my_param_1 - self.param_1_target)
        error_2 = (self.my_param_2 - self.param_2_target)
        
        error = torch.cat((error_1, error_2))
        return error
        

        
```

**Comment by nengqian11** [comment]
I should use [nn.paramlist](https://discuss.pytorch.org/t/model-backpropagation-python-list-vs-nn-parameterlist/148223)..

**Comment by fabiodimarco** [comment]
@nengqian11 I see you figured it out yourself. In PyTorch, plain Python lists won't auto-register parameters, which is why you need to use `nn.ParameterList` (or `nn.ParameterDict` for dictionaries). It ensures all parameters are tracked for optimization, device moves, and state saving.

Have you got any improvement in using LM compared to Adam?

**Comment by nengqian11** [comment]
@fabiodimarco Thanks ! still buggy my code, once I get the result I will see if there are improvenment.



> Hi thanks for your reply! my current code is quite buggy and cannot show you. Here is a very simple example about what I am tring to do.
> 
> ```
> # -----original code, 
> # init some parameters here
> optimizer = torch.optim.Adam(model.parameters(), lr=0.001)
> for epoch in range(max_epochs):
>     optimizer.zero_grad()
>     
>     # some my computaion
>     my_computation()
> 
>     # compute the error here, they have different shape
>     some_error_1 = torch.tensor([10, 6, 3,3])   # just to show the shape
>     some_error_2 = torch.tensor([10, 6, 3])     # just to show the shape
>     some_loss_1 = torch.mean(some_error_1 ** 2)
>     some_loss_2 = torch.mean(some_error_2 ** 2)
>     loss = some_loss_1 + some_loss_2  # a scalar, can directly do backward() update.
>     
>     loss.backward()
>     optimizer.step()
> 
> 
> # ----- try to use the torch LM, so I defined this myModel from nn.Module
> class myModel(nn.Module):
>     def __init__(self) -> None:
>         super().__init__()
>         # some parameters init here
> 
>     def forward(self, _=None) -> torch.Tensor:
>         # some my computaion
>         my_computation()
> 
>         # compute the error here, they have different shape
>         some_error_1 = torch.tensor([10, 6, 3,3])   # just to show the shape
>         some_error_2 = torch.tensor([10, 6, 3])     # just to show the shape
>         some_loss_1 = torch.mean(some_error_1 ** 2)
>         some_loss_2 = torch.mean(some_error_2 ** 2)
>         loss = some_loss_1 + some_loss_2
>         return loss  # this is a scalar
>         # or 
>         # return torch.cat((some_loss_1.view(-1), some_loss_2.view(-1)))  # this is a vector
> ```
> 
> So originally I am using adam to do the optimization, and I do not even have a nn.Module class. I just compute all the error and then add together to the loss scalar, and do backward update. This adam version works fine, I get the correct result, it is just slow.
> 
> So my question is how to actually convert this original code to be optimized by the torch LM optimizer? I tried both way, seems neither is correct. The jacobian I get is all zero.
> 
> ```
>   return loss  # this is a scalar
>         # or 
>   # return torch.cat((some_loss_1.view(-1), some_loss_2.view(-1)))  # this is a vector
> ```

Can I directly return the loss scalar value in my NNModule class's forward in LM solver?  I suppose I cannot, since if I return a scalar, then there will be no Jacobian matrix but a jacobian vector. Then for using LM solver, instead of using a scalar to accumlate all lose, I should seperate them back to each error vector, and cat them?

I am tring to find the way to get the same ( or similar) loss value in LM and adam solver, I guess there will be the issue's regarding to the different loss I am using. In the adam version, most lose are computed by mean root.

**Comment by fabiodimarco** [comment]
@nengqian11 The output of the NNModule can be the loss, but it requires some careful consideration. I do not suggest you do that. 
What type of loss are you using MSE? I suggest you just provide the NNModule  the input data, compute the output and return that. Then use one of the provided loss function as MSE, L1 etc to compare the output of your NNModule with your target data.

Btw, even if your code is buggy you can share it I can help you make it work. Because this way I do not have idea of what you are trying to do.

**Comment by nengqian11** [comment]
@fabiodimarco Thanks,
I am not training a neural network module, instead I am trying to use this LM solver to solve an optimization problem.  It contains some loss which also try to minimize the difference of each parameters. 
Therefore the ouput of my defined NN module is not importent, all I want is try to minimize the total loss to 0.

So I have this original code which is similar to the code block here. And I was using adam to optimize it, the final result is good, but it  is slow. 
And what I am trying to do is try to redo this optimization by your LM solver, and hopefully gain some efficiency improvement and still have same ( or at least similar) result with similar loss value.

So the real problem is like how to simply change the original code to the LM solver?  My original code contains several different loss term ( or energy term in optimization we say), each with different loss function. 
```
# -----original code, 
# init some parameters here
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)
for epoch in range(max_epochs):
    optimizer.zero_grad()
    
    # some my computaion
    my_computation()

    # compute the error here, they have different shape
    some_error_1 = torch.tensor([10, 6, 3,3])   # just to show the shape
    some_error_2 = torch.tensor([10, 6, 3])     # just to show the shape
    some_loss_1 = torch.mean(some_error_1 ** 2)
    some_loss_2 = torch.mean(some_error_2.norm())
    loss = some_loss_1 + some_loss_2  # a scalar, can directly do backward() update.
    
    loss.backward()
    optimizer.step()


# ----- try to use the torch LM, so I defined this myModel from nn.Module
class myModel(nn.Module):
    def __init__(self) -> None:
        super().__init__()
        # some parameters init here

    def forward(self, _=None) -> torch.Tensor:
        # some my computaion
        my_computation()

        # compute the error here, they have different shape
        some_error_1 = torch.tensor([10, 6, 3,3])   # just to show the shape
        some_error_2 = torch.tensor([10, 6, 3])     # just to show the shape
        some_loss_1 = torch.mean(some_error_1 ** 2)
        some_loss_2 = torch.mean(some_error_2 ** 2)
        loss = some_loss_1 + some_loss_2
        return loss  # this is a scalar
        # or 
        # return torch.cat((some_loss_1.view(-1), some_loss_2.view(-1)))  # this is a vector

```

**Comment by fabiodimarco** [comment]
@nengqian11 The problem is that you are doing this operation:
```
        some_loss_1 = torch.mean(some_error_1 ** 2)
        some_loss_2 = torch.mean(some_error_2 ** 2)
        loss = some_loss_1 + some_loss_2
```

While this is ok when using standard first order methods, LM instead will optimize each individual output (in this case 10) using the Jacobian matrix. If you just compute the mean then you will just have a standard first order methods as the Jacobian will have just one row.

In order to properly take advantage of LM you need to return the errors for each example or batch. I think the easiest way could be something like this:

```
        some_error_1 = torch.tensor([10, 6, 3,3])   # just to show the shape
        some_error_2 = torch.tensor([10, 6, 3])     # just to show the shape

        error = torch.cat((some_loss_1.view(10, -1), some_loss_2.view(10, -1)))  # shape: [10, 72]

       return error
```

And then just use MSELoss where the target is set to be just a dummy Tensor of zeros with shape [10, 72].

Another option is to do a larger Jacobian like this:

```
        some_error_1 = torch.tensor([10, 6, 3,3])   # just to show the shape
        some_error_2 = torch.tensor([10, 6, 3])     # just to show the shape

        error = torch.cat((some_loss_1.view(-1), some_loss_2.view(-1))).unsqueeze(1)  # shape: [720, 1]

       return error
```

And then use MSELoss with the target is set to be just a dummy Tensor of zeros with shape [720, 1].

Let me know if this helps.

**Comment by fabiodimarco** [comment]
@nengqian11 any news on that? Can I close the issue?
